core/CacheId.php

```diff
@@ -26,4 +26,112 @@ public static function pluginAware($cacheId)
 
         return $cacheId;
     }
+
+    public static function siteAware($cacheId, array $idSites = null)
+    {
+        if ($idSites === null) {
+            $idSites = self::getIdSiteList('idSite');
+            $cacheId .= self::idSiteListCacheKey($idSites);
+
+            $idSites = self::getIdSiteList('idSites');
+            $cacheId .= self::idSiteListCacheKey($idSites);
+
+            $idSites = self::getIdSiteList('idsite'); // tracker param
+            $cacheId .= self::idSiteListCacheKey($idSites);
+        } else {
+            $cacheId .= self::idSiteListCacheKey($idSites);
+        }
+
+        return $cacheId;
+    }
+
+    private static function getIdSiteList($queryParamName)
+    {
+        if (empty($_GET[$queryParamName])
+            && empty($_POST[$queryParamName])
+        ) {
+            return [];
+        }
+
+        $idSiteGetParam = empty($_GET[$queryParamName]) ? [] : explode(',', $_GET[$queryParamName]);
+        $idSitePostParam = empty($_POST[$queryParamName]) ? [] : explode(',', $_POST[$queryParamName]);
+
+        $idSiteList = array_merge($idSiteGetParam, $idSitePostParam);
+        $idSiteList = array_map('intval', $idSiteList);
+        $idSiteList = array_unique($idSiteList);
+        sort($idSiteList);
+        return $idSiteList;
+    }
+
+    private static function idSiteListCacheKey($idSites)
+    {
+        if (empty($idSites)) {
+            return '';
+        }
+
+        if (count($idSites) <= 5) {
+            return '-' . implode('_', $idSites); // we keep the cache key readable when possible
+        } else {
+            return '-' . md5(implode('_', $idSites)); // we need to shorten it
+        }
+    }
+
+    /**
+     * Temporarily overwrites the idSite parameter so all code executed by `$callback()`
+     * will use cache keys for that idSite.
+     *
+     * Useful when you need to change the idSite context for a chunk of code. For example,
+     * if we are archiving for more than one site in sequence, we don't want to use
+     * the same caches for both archiving executions.
+     *
+     * @param string|int $idSite
+     * @param callable $callback
+     * @return mixed returns result of $callback
+     */
+    public static function overwriteIdSiteForCache($idSite, $callback)
```

Comment:
@benaka sorry to ping here again. It's funny I'm just working on a case where I actually next exactly the same feature. I have a task that iterates over an array of entries that belong to different sites and would need this as well. As this feature is not in any way cache related and can be also used in other places, maybe we move this awesome new feature somewhere else? eg to the `Site` class or so? Any thoughts?

Reply:
Could move it to a `Context` class, like, `Context::changeIdSite()`?

Reply:
works as well 👍